### PR TITLE
Fix hermetic tests in kubeadm

### PR DESCRIPTION
**Pull request by rojkov**
**What type of PR is this?**

/kind cleanup
/kind failing-test

**What this PR does / why we need it**:

The PR fixes the Bazel-based unit test `//cmd/kubeadm/app/cmd:go_default_test` when it's run behind a corporate proxy.

**Special notes for your reviewer**:
When we run `go test` in a pristine git tree checked out to an arbitrary commit then the file `pkg/version/base.go` contains unmodified `gitVersion` variable whose default value ("v0.0.0-master+$Format:%h$") is unparsable.
    
Unfortunately this is the value fed to `kubeadmVersion()` resulting in an error. Even if the error is ignored in the hope that the remote version is used instead then this triggers a test failure in "air-gapped" environment (e.g. behind a corporate proxy)
    
In this special case fall back to the value of k8s.io/kubernetes/cmd/kubeadm/app/constants.MinimumControlPlaneVersion.

/release-note-none
/priority backlog

```release-note
NONE
```

### CODE REVIEW

**Review by neolit123 — COMMENTED** [review]
> When we run go test in a pristine git tree checked out to an arbitrary commit then the file pkg/version/base.go contains unmodified gitVersion variable whose default value ("v0.0.0-master+$Format:%h$") is unparsable.

so at first sight, i don't think we should do this.
i don't know about Bazel but for the regular `make` build if you want the version to be populated correctly from git tags you just have to do:
```
make all WHAT=./cmd/kubeadm
```

the build process writes a version file for this, in the lines of:
```
cat <<EOF > $(pwd)/kube-version
KUBE_GIT_COMMIT=$(git log | head -1 | awk '{print $2}')
KUBE_GIT_TREE_STATE='clean'
KUBE_GIT_VERSION=v1.13.0-alpha.1
KUBE_GIT_MAJOR=1
KUBE_GIT_MINOR=13
EOF
export KUBE_GIT_VERSION_FILE=$(pwd)/kube-version
```
and then uses the above env. variable.

**Comment by rojkov** [comment]
AFAIU Bazel resets environment for the sake of reproducibility. A remote version is used always in CI. But remote versions are inaccessible if behind a proxy.

File: cmd/kubeadm/app/util/version.go [hunk 1]
```diff
@@ -184,7 +187,15 @@ func fetchFromURL(url string, timeout time.Duration) (string, error) {
 func kubeadmVersion(info string) (string, error) {
 	v, err := versionutil.ParseSemantic(info)
 	if err != nil {
-		return "", pkgerrors.Wrap(err, "kubeadm version error")
+		// In case we run `go test` in git tree the version comes from
+		// unmodified pkg/version/base.go. Then we'd better fall back
+		// to a predefined but meaningful constant.
+		if strings.HasPrefix(info, "v0.0.0-master") {
+			klog.Warningf("Can't use in-tree client version. Fall back to minimum control plane version")
+			v = constants.MinimumControlPlaneVersion
```
Comment (neolit123) [inline cmd/kubeadm/app/util/version.go]: it's not a good fallback because it's MINOR-1.

**Comment by neolit123** [comment]
> AFAIU Bazel resets environment for the sake of reproducibility. A remote version is used always in CI. But remote versions are inaccessible if behind a proxy.

there has to be a way to feed the git version into Bazel.
also if Bazel is problematic behind a proxy why not use regular `make` then?

i'm mostly -1 on this change, but deferring to:
/assign @kad
/assign @timothysc 
for now.

**Comment by rojkov** [comment]
@neolit123 Thanks for your review! The PR has revealed other problems I'd like to address first.

File: cmd/kubeadm/app/util/version.go (hunk 1 above)
Comment (rosti) [reply cmd/kubeadm/app/util/version.go]: @rojkov I support the opinion of @neolit123 . We have to be pointing at the current version.
Luckily, @bart0sh is working on a series of PRs, that will introduce some constants, possibly for the current version too. You can sync together for a constant to be used in this PR, but you'll probably have to wait a bit for Ed's PRs.

**Comment by neolit123** [comment]
@rojkov 
both this PR and https://github.com/kubernetes/kubernetes/pull/72059/files
touch the same area of the code.

could please close one of them - preferably this one.
thanks.

**Comment by rojkov** [comment]
@neolit123 sure, closing...
#72059 is ready for review

**Pull request closed without merging**